openhands/runtime/utils/runtime_build.py

```diff
@@ -224,160 +117,181 @@ def build_runtime_image(
+        prep_build_folder(build_folder, base_image, True, extra_deps)
         if not dry_run:
             _build_sandbox_image(
-                docker_folder=cur_docker_build_folder,
-                runtime_builder=runtime_builder,
-                target_image_repo=runtime_image_repo,
-                # NOTE: WE ALWAYS use the "from_scratch_hash" tag for the target image
-                target_image_hash_tag=from_scratch_hash,
-                target_image_tag=runtime_image_tag,
-            )
-        else:
-            logger.info(
-                f'Dry run: Skipping image build for [{generic_runtime_image_name}]'
+                build_folder,
+                runtime_builder,
+                runtime_image_repo,
+                hash_tag,
+                lock_tag,
             )
+        return hash_image_name
+
+    lock_image_name = f'{runtime_image_repo}:{lock_tag}'
+    build_from_scratch = True
+
+    # If the exact image already exists, we do not need to build it
+    if runtime_builder.image_exists(hash_image_name, False):
+        logger.info(f'Reusing Image [{hash_image_name}]')
+        return hash_image_name
+
+    # We look for an existing image that shares the same lock_tag. If such an image exists, we
+    # can use it as the base image for the build and just copy source files. This makes the build
+    # much faster.
+    if runtime_builder.image_exists(lock_image_name):
+        logger.info(f'Build [{hash_image_name}] from [{lock_image_name}]')
+        build_from_scratch = False
+        base_image = lock_image_name
+    else:
+        logger.info(f'Build [{hash_image_name}] from scratch')
+
+    prep_build_folder(build_folder, base_image, build_from_scratch, extra_deps)
+    if not dry_run:
+        _build_sandbox_image(
+            build_folder,
+            runtime_builder,
+            runtime_image_repo,
+            hash_tag,
+            lock_tag,
+        )
 
-        if docker_build_folder is None:
-            shutil.rmtree(cur_docker_build_folder)
-
-    return f'{runtime_image_repo}:{from_scratch_hash}'
+    return hash_image_name
 
 
-def _build_sandbox_image(
-    docker_folder: str,
-    runtime_builder: RuntimeBuilder,
-    target_image_repo: str,
-    target_image_hash_tag: str,
-    target_image_tag: str,
-) -> str:
-    """Build and tag the sandbox image.
-    The image will be tagged as both:
-        - target_image_hash_tag
-        - target_image_tag
+def prep_build_folder(
+    build_folder: Path,
+    base_image: str,
+    build_from_scratch: bool,
+    extra_deps: str | None,
+):
+    # Copy the source code to directory. It will end up in build_folder/code
+    # If package is not found, build from source code
+    openhands_source_dir = Path(openhands.__file__).parent
+    project_root = openhands_source_dir.parent
+    logger.info(f'Building source distribution using project root: {project_root}')
+
+    # Copy the 'openhands' directory (Source code)
+    shutil.copytree(
```

Comment:
I still kinda prefer `sdist` over `shutil.copytree` as long as it is not TOO slow.

The thing is: user may unexpectly put other files in the source directly (e.g., a JSON that's extremely large) -- although this is not developing best practices, but it could still happen (sometime may happen with me if I'm debugging stuff).

We could only copy over `*.py` but not sure if that will cause problems - using `sdist` is probably the safest here.

Reply:
I don't see much difference in performance - but the change to use shutil was done last week by the OS community: (It looks like sdist does not work when OpenHands is being used as a python library) https://github.com/All-Hands-AI/OpenHands/commit/568c8ce9936f83feaab87bd6ee34d2f39efbc0e6